File: sensors/other/aes.py
#import hashlib
import binascii
import ucryptolib
import os


class AES():

    def __init__(self, key): 
        self.bs = 16
        self.key = binascii.unhexlify(key) # key must be a hex string

    def encrypt(self, plainText):
        plainText = self._pad(plainText)
        iv = os.urandom(self.bs)
        cipher = ucryptolib.aes(self.key, 2,iv) #CBC mode
        cipherText=binascii.b2a_base64(cipher.encrypt(plainText)).decode('utf-8')
        return binascii.b2a_base64(iv).decode("utf-8")+":"+cipherText

    def decrypt(self, cipherTextWithIV):
        iv,cipherText=cipherTextWithIV.split(":")
        cipherText = binascii.a2b_base64(cipherText)
        iv=binascii.a2b_base64(iv)
        cipher = ucryptolib.aes(self.key, 2,iv)
        return self._unpad(cipher.decrypt(cipherText)).decode('utf-8')

    def _pad(self, s):
        return s + (self.bs - len(s) % self.bs) * chr(self.bs - len(s) % self.bs)

    def _unpad(self,s):
        return s[:-ord(s[len(s)-1:])]


""" class AES():

    def __init__(self, key): 
        self.bs = 16
        self.key = binascii.unhexlify(key) # key must be a hex string

    def encrypt(self, plainText):
        plainText = self._pad(plainText)
        iv = os.urandom(self.bs)
        cipher = ucryptolib.aes(self.key, 2,iv) #CBC mode
        cipherText=binascii.hexlify(cipher.encrypt(plainText)).decode('utf-8')
        return binascii.hexlify(iv).decode("utf-8")+":"+cipherText

    def decrypt(self, cipherTextWithIV):
        iv,cipherText=cipherTextWithIV.split(":")
        cipherText = binascii.unhexlify(cipherText)
        iv=binascii.unhexlify(iv)
        cipher = ucryptolib.aes(self.key, 2,iv)
        return self._unpad(cipher.decrypt(cipherText)).decode('utf-8')

    def _pad(self, s):
        return s + (self.bs - len(s) % self.bs) * chr(self.bs - len(s) % self.bs)

    def _unpad(self,s):
        return s[:-ord(s[len(s)-1:])]
 """

File: server/src/security/aes.js
const Crypto = require('crypto');


class AES128CBC {

    constructor(key) {
        this.key = Buffer.from(key, "hex");
        this.bs = 16;
    }

    decrypt(cipherTextWithIV) {

        let cipherText, iv;
        let plainText = "";

        [iv, cipherText] = cipherTextWithIV.split(":"); // split iv and ciphertext
        iv = Buffer.from(iv, "base64");

        const decipher = Crypto.createDecipheriv("aes-128-cbc", this.key, iv);

        plainText = decipher.update(cipherText, 'base64', 'utf8');
        plainText += decipher.final('utf8');

        return plainText;

    }

    encrypt(plainText) {

        let cipherText = "";

        let iv = Crypto.randomBytes(this.bs)

        const cipher = Crypto.createCipheriv('aes-128-cbc', this.key, iv);

        cipherText = cipher.update(plainText, 'utf8', 'base64');
        cipherText += cipher.final('base64');

        cipherText = iv.toString("base64") + ":" + cipherText;

        return cipherText;
    }


}

module.exports = {
    /**
     * Creates an AES128CBC Cipher object.
     *
     * @param {String} key
     *   Key for cipher. {key} Must be a hex string
     */
    createAES128Cipher(key) {
        return new AES128CBC(key);
    }
};